platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/LongList.java

```diff
@@ -121,31 +149,53 @@ protected LongList(final FileChannel fileChannel) throws IOException {
         this.fileChannel = fileChannel;
         if (fileChannel.size() > 0) {
             // read header from existing file
-            final ByteBuffer headerBuffer = ByteBuffer.allocate(FILE_HEADER_SIZE);
-            MerkleDbFileUtils.completelyRead(fileChannel, headerBuffer);
-            headerBuffer.rewind();
-            final int formatVersion = headerBuffer.getInt();
-            if (formatVersion != FILE_FORMAT_VERSION) {
-                throw new IOException("Tried to read a file with incompatible file format version ["
+            final ByteBuffer versionBuffer = readFromFileChanel(fileChannel, VERSION_METADATA_SIZE);
+            final int formatVersion = versionBuffer.getInt();
+            final int formatMetadataSize;
+            if (formatVersion == INITIAL_VERSION) {
+                formatMetadataSize = FORMAT_METADATA_SIZE_V1;
+                currentFileHeaderSize = FILE_HEADER_SIZE_V1;
+            } else if (formatVersion == MIN_VALID_INDEX_SUPPORT_VERSION) {
+                formatMetadataSize = FORMAT_METADATA_SIZE_V2;
+                currentFileHeaderSize = FILE_HEADER_SIZE_V2;
+            } else {
+                throw new IOException("File format version is not supported. File format version ["
                         + formatVersion
-                        + "], expected ["
-                        + FILE_FORMAT_VERSION
+                        + "], the latest supported version is ["
+                        + CURRENT_FILE_FORMAT_VERSION
                         + "].");
             }
+
+            final ByteBuffer headerBuffer = readFromFileChanel(fileChannel, formatMetadataSize);
             numLongsPerChunk = headerBuffer.getInt();
             memoryChunkSize = numLongsPerChunk * Long.BYTES;
             maxIndexThatCanBeStored.set(headerBuffer.getLong());
             maxLongs = headerBuffer.getLong();
-            size.set((fileChannel.size() - FILE_HEADER_SIZE) / Long.BYTES);
+            if (formatVersion > INITIAL_VERSION) {
+                minValidIndex.set(headerBuffer.getLong());
+                // "inflating" the size by number of indices that are to the left of the min valid index
+                size.set(minValidIndex.get() + (fileChannel.size() - currentFileHeaderSize) / Long.BYTES);
+            } else {
+                size.set((fileChannel.size() - FILE_HEADER_SIZE_V1) / Long.BYTES);
+            }
         } else {
             // opening a new file
             numLongsPerChunk = DEFAULT_NUM_LONGS_PER_CHUNK;
             memoryChunkSize = numLongsPerChunk * Long.BYTES;
             maxLongs = DEFAULT_MAX_LONGS_TO_STORE;
+            currentFileHeaderSize = FILE_HEADER_SIZE_V2;
             writeHeader(fileChannel);
         }
     }
 
+    private static ByteBuffer readFromFileChanel(final FileChannel fileChannel, final int bytesToRead)
```

Comment:
typo: `readFromFileChanel` ==> `readFromFileChannel`

Reply:
Thank you for catching this.